File: Algorithm/LongestSubstringWithoutRepeatingCharacters.java
// Using a hashtable to remember the last index of every char. And keep tracking the starting point of a valid substring.
// start = max(start, last[s[i]] + 1)
// ans = max(ans, i - start + 1)
// Time complexity: O(n)
// Space complexity: O(128)
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int[] last = new int[128];
        Arrays.fill(last, -1);
        int start = 0;
        int ans = 0;
        for (int i = 0; i < s.length(); ++i) {
            if (last[s.charAt(i)] != -1) 
                start = Math.max(start, last[s.charAt(i)] + 1);
            last[s.charAt(i)] = i;
            ans = Math.max(ans, i - start + 1);
        }
        return ans;
    }
}

File: Algorithm/LongestSubstringWithoutRepeatingCharacters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        const int n = s.length();
        int ans = 0;
        vector<int> idx(128, -1);
        for (int i = 0, j = 0; j < n; ++j) {
            i = max(i, idx[s[j]] + 1);
            ans = max(ans, j - i + 1);
            idx[s[j]] = j;
        }
        return ans;
    }
};
